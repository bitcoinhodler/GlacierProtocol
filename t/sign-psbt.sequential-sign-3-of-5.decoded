{
  "tx": {
    "txid": "932f37961f4c73f5344b5018c32638b5a8977b78bbc56bb672b293b35b723469",
    "hash": "932f37961f4c73f5344b5018c32638b5a8977b78bbc56bb672b293b35b723469",
    "version": 2,
    "size": 85,
    "vsize": 85,
    "weight": 340,
    "locktime": 0,
    "vin": [
      {
        "txid": "1a9a660e8e48f1688bb8fc613a3c2761c43f5879eecf11d59415d4c5349415c3",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 0.19997810,
        "n": 0,
        "scriptPubKey": {
          "asm": "OP_DUP OP_HASH160 b6c7733e23a271e57ca16e5cd885174917968db9 OP_EQUALVERIFY OP_CHECKSIG",
          "desc": "addr(mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99)#fxf0ktmw",
          "hex": "76a914b6c7733e23a271e57ca16e5cd885174917968db988ac",
          "address": "mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99",
          "type": "pubkeyhash"
        }
      }
    ]
  },
  "global_xpubs": [
  ],
  "psbt_version": 0,
  "proprietary": [
  ],
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 0.20000000,
        "scriptPubKey": {
          "asm": "OP_HASH160 2524a7e29329a636bf4c1d8dea0dc6a087e5d916 OP_EQUAL",
          "desc": "addr(2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS)#jg2pk0wd",
          "hex": "a9142524a7e29329a636bf4c1d8dea0dc6a087e5d91687",
          "address": "2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS",
          "type": "scripthash"
        }
      },
      "partial_signatures": {
        "0320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced": "304402201d19a0b425765eff8289f7a3df7d8fc4781cdcc064424041ff1f5c5dbe95145302200c3da2466a186b8d14d693c63a364066b7463fd0dc9c1d7d151786d9e08ec07f01"
      },
      "redeem_script": {
        "asm": "0 8e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e",
        "hex": "00208e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e",
        "type": "witness_v0_scripthash"
      },
      "witness_script": {
        "asm": "3 0320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced 02766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e9 039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d6 03fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e01 029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee4183151 5 OP_CHECKMULTISIG",
        "hex": "53210320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced2102766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e921039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d62103fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e0121029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee418315155ae",
        "type": "multisig"
      }
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.00002190
}

{
  "tx": {
    "txid": "932f37961f4c73f5344b5018c32638b5a8977b78bbc56bb672b293b35b723469",
    "hash": "932f37961f4c73f5344b5018c32638b5a8977b78bbc56bb672b293b35b723469",
    "version": 2,
    "size": 85,
    "vsize": 85,
    "weight": 340,
    "locktime": 0,
    "vin": [
      {
        "txid": "1a9a660e8e48f1688bb8fc613a3c2761c43f5879eecf11d59415d4c5349415c3",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967295
      }
    ],
    "vout": [
      {
        "value": 0.19997810,
        "n": 0,
        "scriptPubKey": {
          "asm": "OP_DUP OP_HASH160 b6c7733e23a271e57ca16e5cd885174917968db9 OP_EQUALVERIFY OP_CHECKSIG",
          "desc": "addr(mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99)#fxf0ktmw",
          "hex": "76a914b6c7733e23a271e57ca16e5cd885174917968db988ac",
          "address": "mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99",
          "type": "pubkeyhash"
        }
      }
    ]
  },
  "global_xpubs": [
  ],
  "psbt_version": 0,
  "proprietary": [
  ],
  "unknown": {
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 0.20000000,
        "scriptPubKey": {
          "asm": "OP_HASH160 2524a7e29329a636bf4c1d8dea0dc6a087e5d916 OP_EQUAL",
          "desc": "addr(2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS)#jg2pk0wd",
          "hex": "a9142524a7e29329a636bf4c1d8dea0dc6a087e5d91687",
          "address": "2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS",
          "type": "scripthash"
        }
      },
      "partial_signatures": {
        "0320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced": "304402201d19a0b425765eff8289f7a3df7d8fc4781cdcc064424041ff1f5c5dbe95145302200c3da2466a186b8d14d693c63a364066b7463fd0dc9c1d7d151786d9e08ec07f01",
        "03fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e01": "304402205c3406119f6f7b1bf72d32c11d3a5fe974a27c5102345fcc5ed7af61e647f70202203e4e2a9496e51e0b9d61e2262c1b16e144f9827fb644a40baae212e9fadd1a1d01"
      },
      "redeem_script": {
        "asm": "0 8e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e",
        "hex": "00208e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e",
        "type": "witness_v0_scripthash"
      },
      "witness_script": {
        "asm": "3 0320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced 02766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e9 039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d6 03fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e01 029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee4183151 5 OP_CHECKMULTISIG",
        "hex": "53210320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced2102766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e921039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d62103fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e0121029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee418315155ae",
        "type": "multisig"
      }
    }
  ],
  "outputs": [
    {
    }
  ],
  "fee": 0.00002190
}
{
  "txid": "e15dc18bea9e2739badf188e85b766dafaf881c8512a4d63001280029f4004d4",
  "hash": "e2b8ac5021a57fb93a74e29a9ee02ad15fc4953c84a817244a9c2f5e9d698f89",
  "version": 2,
  "size": 514,
  "vsize": 219,
  "weight": 874,
  "locktime": 0,
  "vin": [
    {
      "txid": "1a9a660e8e48f1688bb8fc613a3c2761c43f5879eecf11d59415d4c5349415c3",
      "vout": 1,
      "scriptSig": {
        "asm": "00208e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e",
        "hex": "2200208e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5e"
      },
      "txinwitness": [
        "",
        "304402201d19a0b425765eff8289f7a3df7d8fc4781cdcc064424041ff1f5c5dbe95145302200c3da2466a186b8d14d693c63a364066b7463fd0dc9c1d7d151786d9e08ec07f01",
        "304402205c3406119f6f7b1bf72d32c11d3a5fe974a27c5102345fcc5ed7af61e647f70202203e4e2a9496e51e0b9d61e2262c1b16e144f9827fb644a40baae212e9fadd1a1d01",
        "304402204cdfb7a04ca28244c1488e9581ca109d187410be2c1f803fe2afc4ec05cf1734022010e347e665ea45c31aab50ba211f67fca1d10565be5368ff4cda96e96fb6a93b01",
        "53210320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced2102766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e921039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d62103fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e0121029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee418315155ae"
      ],
      "sequence": 4294967295
    }
  ],
  "vout": [
    {
      "value": 0.19997810,
      "n": 0,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 b6c7733e23a271e57ca16e5cd885174917968db9 OP_EQUALVERIFY OP_CHECKSIG",
        "desc": "addr(mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99)#fxf0ktmw",
        "hex": "76a914b6c7733e23a271e57ca16e5cd885174917968db988ac",
        "address": "mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99",
        "type": "pubkeyhash"
      }
    }
  ]
}
