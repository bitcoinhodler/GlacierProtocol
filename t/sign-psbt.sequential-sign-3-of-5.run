#!/bin/bash
set -e

# Test 3-step serial sequential signing, where PSBT created by first
# GlacierScript run is fed to second run, and that PSBT is fed to
# third run.

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
../../t/sign-psbt.sequential-sign-3-of-5.psbt
y
1
cN5pwP2oCxokXYFdRVYczUqLqdA8GYqGWfRjaprrQXgHrS6PTHaD
y
INPUT

# Ugly: need to delete wallet, otherwise the second signing will still have
# the first key and will create two signatures.
echo First glacierscript run complete. Deleting wallet...

# Decode the generated transaction-0*.pngs to get the PSBT for the
# second signature.
zbarimg --set "*.enable=0" --set qr.enable=1 --quiet --raw transaction.png > part1-out.psbt

# Combine this (possibly stripped-down) partial-sig PSBT with the original PSBT
# to get a partial-sig PSBT that GlacierScript can then complete.
bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 combinepsbt '["'`cat ../../t/sign-psbt.sequential-sign-3-of-5.psbt`'", "'`cat part1-out.psbt`'"]' > part2-in.psbt


bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 stop
# So ugly, but without this, I get timeouts restarting bitcoind:
sleep 5
rm -rf ../bitcoin-data/$1/regtest/wallets

echo Running glacierscript again for the second signature...

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
part2-in.psbt
y
1
cVyhd5Ei9HX2wSny8S2nYYtW2ubiUPhzdx1JUiSTjNicQJYTGLQV
y
INPUT


echo Second glacierscript run complete. Deleting wallet...

# Decode the generated transaction-0*.pngs to get the PSBT for the
# second signature.
zbarimg --set "*.enable=0" --set qr.enable=1 --quiet --raw transaction.png > part2-out.psbt

# Combine this (possibly stripped-down) partial-sig PSBT with the original PSBT
# to get a partial-sig PSBT that GlacierScript can then complete.
bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 combinepsbt '["'`cat ../../t/sign-psbt.sequential-sign-3-of-5.psbt`'", "'`cat part2-out.psbt`'"]' > part3-in.psbt


bitcoin-cli -regtest -rpcport=$1 -datadir=../bitcoin-data/$1 stop
# So ugly, but without this, I get timeouts restarting bitcoind:
sleep 5
rm -rf ../bitcoin-data/$1/regtest/wallets

echo Running glacierscript again for the third signature...

$GLACIERSCRIPT --regtest=$1 sign-psbt << INPUT 2>&1
y
y
y
y
y
y
part3-in.psbt
y
1
cUv81BVz1JWuA9uwECPvfsoyA1TqmFJBJaQUPYrxkdf44zQSiTZY
y
INPUT
