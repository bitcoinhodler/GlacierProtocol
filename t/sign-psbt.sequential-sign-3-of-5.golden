Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 3 required to complete transaction)?
#: Output will be a partially signed bitcoin transaction (PSBT).
Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190

Signing with private keys: 
cN5pwP2oCxokXYFdRVYczUqLqdA8GYqGWfRjaprrQXgHrS6PTHaD


Confirm? (y/n): 
Calculating transaction...

Incomplete PSBT (base64):
cHNidP8BAFUCAAAAAcMVlDTF1BWU1RHP7nlYP8RhJzw6Yfy4i2jxSI4OZpoaAQAAAAD/////AXIkMQEAAAAAGXapFLbHcz4jonHlfKFuXNiFF0kXlo25iKwAAAAAAAEBIAAtMQEAAAAAF6kUJSSn4pMppja/TB2N6g3GoIfl2RaHIgIDIISWB+7SZcCfUc2g6XJCZVgnY4SHBlElCqToY4wzTO1HMEQCIB0ZoLQldl7/gon3o999j8R4HNzAZEJAQf8fXF2+lRRTAiAMPaJGahhrjRTWk8Y6NkBmt0Y/0NycHX0VF4bZ4I7AfwEBBCIAII4DvWN+v+wEnRnfeNkAhRQerL0UA0eTFV84Iv1Y09xeAQWtUyEDIISWB+7SZcCfUc2g6XJCZVgnY4SHBlElCqToY4wzTO0hAnZscEcqij3nUdEm7drIQ2M4+ER7GWe1qjsSormm5xLpIQOQgN3ESOf+SYVnan0tZWffwv67ggem8puXNyM/TDtE1iED+i0sB2U6+wpzNA7p5rEP8NRiTJOwESfppp777UhGjgEhApZpt51+BfIMp6pBjtyDnakERnJTyeRxlMuD5p7kGDFRVa4AAA==

Transaction fingerprint (md5):
f3a6e5ed330c3b5d8b242ac7629f4a5b
QR code for transaction written to transaction.png
First glacierscript run complete. Deleting wallet...
Bitcoin Core stopping
Running glacierscript again for the second signature...
Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 2 required to complete transaction)?
#: Output will be a partially signed bitcoin transaction (PSBT).
Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190

Signing with private keys: 
cVyhd5Ei9HX2wSny8S2nYYtW2ubiUPhzdx1JUiSTjNicQJYTGLQV


Confirm? (y/n): 
Calculating transaction...

Incomplete PSBT (base64):
cHNidP8BAFUCAAAAAcMVlDTF1BWU1RHP7nlYP8RhJzw6Yfy4i2jxSI4OZpoaAQAAAAD/////AXIkMQEAAAAAGXapFLbHcz4jonHlfKFuXNiFF0kXlo25iKwAAAAAAAEBIAAtMQEAAAAAF6kUJSSn4pMppja/TB2N6g3GoIfl2RaHIgIDIISWB+7SZcCfUc2g6XJCZVgnY4SHBlElCqToY4wzTO1HMEQCIB0ZoLQldl7/gon3o999j8R4HNzAZEJAQf8fXF2+lRRTAiAMPaJGahhrjRTWk8Y6NkBmt0Y/0NycHX0VF4bZ4I7AfwEiAgP6LSwHZTr7CnM0DunmsQ/w1GJMk7ARJ+mmnvvtSEaOAUcwRAIgXDQGEZ9vexv3LTLBHTpf6XSifFECNF/MXtevYeZH9wICID5OKpSW5R4LnWHiJiwbFuFE+YJ/tkSkC6riEun63RodAQEEIgAgjgO9Y36/7ASdGd942QCFFB6svRQDR5MVXzgi/VjT3F4BBa1TIQMghJYH7tJlwJ9RzaDpckJlWCdjhIcGUSUKpOhjjDNM7SECdmxwRyqKPedR0Sbt2shDYzj4RHsZZ7WqOxKiuabnEukhA5CA3cRI5/5JhWdqfS1lZ9/C/ruCB6bym5c3Iz9MO0TWIQP6LSwHZTr7CnM0DunmsQ/w1GJMk7ARJ+mmnvvtSEaOASEClmm3nX4F8gynqkGO3IOdqQRGclPJ5HGUy4PmnuQYMVFVrgAA

Transaction fingerprint (md5):
edc24b89bb0d0e1e670389445183a84a
QR code for transaction written to transaction.png
Second glacierscript run complete. Deleting wallet...
Bitcoin Core stopping
Running glacierscript again for the third signature...
Are you running this on a computer WITHOUT a network connection of any kind? (y/n)?Have the wireless cards in this computer been physically removed? (y/n)?Are you running on battery power? (y/n)?Are you running on an operating system booted from a USB drive? (y/n)?Is your screen hidden from view of windows, cameras, and other people? (y/n)?Are smartphones and all other nearby devices turned off and in a Faraday bag? (y/n)?Input a filename located in the current directory which contains the PSBT:
0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190
Confirm? (y/n): 
How many private keys will you be signing this transaction with (at least 1 required to complete transaction)?
#: Key #1: 
Is this data correct?
*** WARNING: Incorrect data may lead to loss of funds ***

0.20000000 BTC in unspent inputs from cold storage address 2MvdcuGhBWhwiHjTr7XUSydtv5FmMge93jS
0.19997810 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
Fee amount: 0.00002190

Signing with private keys: 
cUv81BVz1JWuA9uwECPvfsoyA1TqmFJBJaQUPYrxkdf44zQSiTZY


Confirm? (y/n): 
Calculating transaction...

Final fee rate: 10.00 satoshis per vbyte

Raw signed transaction (hex):
02000000000101c3159434c5d41594d511cfee79583fc461273c3a61fcb88b68f1488e0e669a1a01000000232200208e03bd637ebfec049d19df78d90085141eacbd14034793155f3822fd58d3dc5effffffff0172243101000000001976a914b6c7733e23a271e57ca16e5cd885174917968db988ac050047304402201d19a0b425765eff8289f7a3df7d8fc4781cdcc064424041ff1f5c5dbe95145302200c3da2466a186b8d14d693c63a364066b7463fd0dc9c1d7d151786d9e08ec07f0147304402205c3406119f6f7b1bf72d32c11d3a5fe974a27c5102345fcc5ed7af61e647f70202203e4e2a9496e51e0b9d61e2262c1b16e144f9827fb644a40baae212e9fadd1a1d0147304402204cdfb7a04ca28244c1488e9581ca109d187410be2c1f803fe2afc4ec05cf1734022010e347e665ea45c31aab50ba211f67fca1d10565be5368ff4cda96e96fb6a93b01ad53210320849607eed265c09f51cda0e972426558276384870651250aa4e8638c334ced2102766c70472a8a3de751d126eddac8436338f8447b1967b5aa3b12a2b9a6e712e921039080ddc448e7fe4985676a7d2d6567dfc2febb8207a6f29b9737233f4c3b44d62103fa2d2c07653afb0a73340ee9e6b10ff0d4624c93b01127e9a69efbed48468e0121029669b79d7e05f20ca7aa418edc839da904467253c9e47194cb83e69ee418315155ae00000000

Transaction fingerprint (md5):
41012a83b360488722ba8681e1f1587c
QR code for transaction written to transaction.png
